### PR TITLE
feat: Create the pose divergence output port for RedundantRBSSelectorTask

### PR DESCRIPTION
**Pull request by DeboraTahara**

### CODE REVIEW

File: tasks/RedundantRBSSelectorTask.cpp [hunk 1]
```diff
@@ -193,7 +202,24 @@ void RedundantRBSSelectorTask::updateHook()
...
+    _pose_deviation.write(pose_deviation);
 }
+
+bool RedundantRBSSelectorTask::arePosesIncompatible(RigidBodyState const& first,
+    RigidBodyState const& second)
+{
+    base::Position first_position = first.getPose().position;
+    base::Position second_position = second.getPose().position;
+    bool position_status = (first_position - second_position).norm() > m_position_threshold;
+    bool roll_status = isDifferenceOutsideThreshold(first.getRoll(), second.getRoll(), m_roll_threshold);
```
Comment (seixasxbr) [inline tasks/RedundantRBSSelectorTask.cpp]: Imagine a threshold of 30 degrees, the first equal to 170 degrees and the second -170. This difference should be inside the threshold, but in the actual implementation it will be outside -> 170 -(-170) = 340

**Review by seixasxbr — CHANGES_REQUESTED** [review]

File: transforms.orogen [hunk 2]
```diff
@@ -87,11 +87,17 @@ task_context "RedundantRBSSelectorTask" do
 
     property "source_timeout", "/base/Time"
     property "main_source_histeresys", "/base/Time"
+    property "position_threshold", "double"
+    property "roll_threshold", "double"
+    property "pitch_threshold", "double"
+    property "yaw_threshold", "double"
```
Comment (seixasxbr) [inline transforms.orogen:112]: typekit export transforms/PoseDeviation

**Comment by DeboraTahara** [comment]
ping @doudou

File: tasks/RedundantRBSSelectorTask.cpp [hunk 3]
```diff
@@ -194,6 +208,29 @@ void RedundantRBSSelectorTask::updateHook()
         _rbs_out.write(secondary_rbs);
     }
 }
+
+bool RedundantRBSSelectorTask::areOrientationDivergent(Orientation first,
+    Orientation second)
+{
+    Vector3d diff_orientation = getEuler(first) - getEuler(second);
```
Comment (seixasxbr) [inline tasks/RedundantRBSSelectorTask.cpp]: This not seen to be right. Test an edge case (3.14, -3.14)
I would recommend to create a base::Angle from the euler angles, and use the - operator already implemented.

File: tasks/RedundantRBSSelectorTask.cpp [hunk 4]
```diff
@@ -194,6 +206,31 @@ void RedundantRBSSelectorTask::updateHook()
         _rbs_out.write(secondary_rbs);
     }
 }
+
+PoseDivergence RedundantRBSSelectorTask::checkDivergences(
```
Comment (seixasxbr) [inline tasks/RedundantRBSSelectorTask.cpp:205]: To optimize this method, we should use getEuler instead of getRoll, getPich and getYaw.
I forgot to do that when I implemented this

File: tasks/RedundantRBSSelectorTask.cpp [hunk 5]
```diff
@@ -142,17 +149,22 @@ void RedundantRBSSelectorTask::updateHook()
...
     }
 
     bool secondary_is_new_and_valid{false};
-    base::samples::RigidBodyState secondary_rbs;
+    samples::RigidBodyState secondary_rbs;
     if (_secondary_rbs_source.read(secondary_rbs) == RTT::NewData) {
         secondary_is_new_and_valid = updateSecondarySource(secondary_rbs);
     }
 
+    if (main_is_new_and_valid && secondary_is_new_and_valid) {
```
Comment (doudou) [inline tasks/RedundantRBSSelectorTask.cpp]: `||`

The probably that both will be new and valid at the same time is almost zero

File: tasks/RedundantRBSSelectorTask.cpp [hunk 6]
```diff
@@ -194,6 +206,31 @@ void RedundantRBSSelectorTask::updateHook()
...
+    divergence.time = base::Time::now();
+    divergence.position_error_norm = (first.position - second.position).norm();
+    divergence.position_divergent =
+        divergence.position_error_norm > m_position_error_threshold;
+    divergence.roll_error =
+        base::Angle::fromRad(first.getRoll()) - base::Angle::fromRad(second.getRoll());
+    divergence.pitch_error =
+        base::Angle::fromRad(first.getPitch()) - base::Angle::fromRad(second.getPitch());
+    divergence.yaw_error =
+        base::Angle::fromRad(first.getYaw()) - base::Angle::fromRad(second.getYaw());
```
Comment (doudou) [inline tasks/RedundantRBSSelectorTask.cpp]: These are strictly equivalent to the less noisy `base::Angle::fromRad(first.getRoll() - second.getRoll())`

Also, I find it ... interesting that this PR dutifully removes `base::` from RigidBodyState but starts using here for Angle

File: tasks/RedundantRBSSelectorTask.cpp [hunk 7]
```diff
@@ -194,6 +206,31 @@ void RedundantRBSSelectorTask::updateHook()
...
+    divergence.roll_divergent =
+        std::fabs(divergence.roll_error.getRad()) > m_angles_error_thresholds.roll;
+    divergence.pitch_divergent =
+        std::fabs(divergence.pitch_error.getRad()) > m_angles_error_thresholds.pitch;
+    divergence.yaw_divergent =
+        std::fabs(divergence.yaw_error.getRad()) > m_angles_error_thresholds.yaw;
```
Comment (doudou) [inline tasks/RedundantRBSSelectorTask.cpp]: This is ripe for an `evaluateAngleDivergence(base::Angle const& first, base::Angle const& second) -> pair<base::Angle, bool>`

Hint: look at tie https://en.cppreference.com/w/cpp/utility/tuple/tie

File: test/redundant_rbs_selector_task_test.rb [hunk 8]
```diff
@@ -42,6 +46,177 @@
...
+                main_w.write main_rbs
+                secondary_w.write secondary_rbs
+            end.to do
+                have_one_new_sample(task.pose_divergence_port)
+            end
+            expected = { roll_divergent: 1, roll_error: 0.7 }
+            assert_divergence(result, expected)
+        end
+
+        it "returns false when the roll difference is smaller than its threshold" do
```
Comment (doudou) [inline test/redundant_rbs_selector_task_test.rb]: This is a component. It does not "return false". `it "does not return a divergence in roll if the roll difference is smaller than the configured threshold"`

File: transformsTypes.hpp [hunk 9]
```diff
@@ -9,7 +13,22 @@
...
+        bool yaw_divergent = false;
+        double position_error_norm = base::unknown<double>();
+        base::Angle roll_error;
+        base::Angle yaw_error;
+        base::Angle pitch_error;
+    };
+    struct AngleErrorThresholds {
+        double roll = base::unknown<double>();
+        double pitch = base::unknown<double>();
+        double yaw = base::unknown<double>();
```
Comment (doudou) [inline transformsTypes.hpp]: ... sigh ... base::Angle

File: transformsTypes.hpp [hunk 10]
```diff
@@ -9,7 +13,22 @@
  */
 
 namespace transforms {
+    struct PoseDivergence {
```
Comment (doudou) [inline transformsTypes.hpp:16]: I know that I always say that optimization is secondary, but I'll make an exception in this case.

errors are most probably small, so would make sense to stick to float for position_error_norm

Which is actually going to save 8 bytes per struct. The `bool` are one byte each but then the compiler is adding 4 bytes between yaw_divergent and position_error_norm for alignment reasons.

File: transformsTypes.hpp [hunk 11]
```diff
@@ -9,7 +13,22 @@
...
+        base::Time time;
+        bool position_divergent = false;
+        bool roll_divergent = false;
+        bool pitch_divergent = false;
+        bool yaw_divergent = false;
+        double position_error_norm = base::unknown<double>();
```
Comment (doudou) [inline transformsTypes.hpp]: Why is it `position_error_norm` but we don't have anything for the angles ?

I think that `position_error` would be more than enough, and we win for consistency.

File: tasks/RedundantRBSSelectorTask.cpp (hunk 7 above)
Comment (DeboraTahara) [reply tasks/RedundantRBSSelectorTask.cpp]: I changed it to use the getEuler() instead of get roll, pitch, yaw separated. Since I did the difference between to vectors of eulers, I think its okay not to have this. If you still feel it's not good, I can change it

File: transforms.orogen [hunk 12]
```diff
@@ -8,6 +8,8 @@ import_types_from "transformsTypes.hpp"
 import_types_from "std"
 import_types_from "base"
 
+typekit.export_types "/transforms/PoseDivergence"
+
```
Comment (doudou) [inline transforms.orogen]: What is that doing here ?

File: tasks/RedundantRBSSelectorTask.cpp [hunk 13]
```diff
@@ -50,6 +53,11 @@ bool isSecondarySourceValid(RedundantRBSSelectorTask::InternalState const& state
     return base::Time::now() < state.secondary_source_deadline;
 }
 
+bool isDifferenceOutsideThreshold(double first, double second, double threshold)
+{
+    return std::abs(first - second) > threshold;
+}
+
```
Comment (doudou) [inline tasks/RedundantRBSSelectorTask.cpp]: Not in use anymore.

**Review by doudou — APPROVED** [review]
Just fix the two remaining comments and it's good to go from my point of view.

**Review by seixasxbr — APPROVED** [review]
